Выполнил домашнюю работу №1

MAMKUH6OPIIIUK opened the pull request.

MAMKUH6OPIIIUK committed:
Выполнил домашнюю работу №1

MAMKUH6OPIIIUK committed:
Добавил try-with-resources при получении InputStream с csv-файлом

vlad-iv commented on spring-hw-01/pom.xml:
```diff
@@ -0,0 +1,113 @@
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <spring.version>6.0.6</spring.version>
+        <opencsv.version>4.1</opencsv.version>
+        <junit.version>4.13</junit.version>
```
Предлагаю обновить до JUnit 5 версии

MAMKUH6OPIIIUK replied:
Поднял до версии 5.9.2, добавил конфигурацию maven-surefire-plugin (иначе при сборке моки не инжектились), ну и переписал сам тестовый класс, т.к. некоторые использовавшиеся аннотации\методы отсутствуют в JUnit 5

vlad-iv commented on spring-hw-01/src/main/java/ru/otus/spring/homework/oke/Main.java:
```diff
@@ -0,0 +1,13 @@
+    public static void main(String[] args) {
+        ClassPathXmlApplicationContext context = new ClassPathXmlApplicationContext("/spring-context.xml");
+        TestingService testingService = context.getBean(TestingService.class);
+        testingService.executeStudentTesting("Java testing from CSV", 40);
```
Лучше использовать метод без параметров. 
Если они требуются можно задать при конфигурировании бина TestingService

MAMKUH6OPIIIUK replied:
Переделал бин TestingService - теперь тема, по которой должен проверяться студент, и проходной балл по этой теме являются полями этого класса. Метод тестирования студента теперь не требует аргументов, а использует значения полей.
Переделал конфигурацию бина TestingService - теперь конфигурируется через конструктор, а не через setter.

vlad-iv commented on spring-hw-01/src/main/java/ru/otus/spring/homework/oke/dao/TestingDaoCsv.java:
```diff
@@ -0,0 +1,152 @@
+ * данные возвращаются именно из этого кэша
+ */
+public class TestingDaoCsv implements TestingDao {
+    private final Testing testingCache;
```
Кеширование - хорошая оптимизация. В Spring за это отвечают другие бины (мы будем проходить)
Предлагаю отключить кеширование.

MAMKUH6OPIIIUK replied:
Спасибо! 👍 Я догадывался, что что-то такое в Spring быть должно. Убрал запись в поле класса `private final Testing testingCache` (как и само поле). Теперь при каждом запуске findAll читается ресурс и из него конструируется новый список тестирований

vlad-iv commented on spring-hw-01/src/main/java/ru/otus/spring/homework/oke/dao/TestingDaoCsv.java:
```diff
@@ -0,0 +1,152 @@
+     *                                     См. {@link #parseLineAndSave(String[])}
+     */
+    public TestingDaoCsv(String csvResourceName) throws IOException, IncorrectCsvFormatException {
+        this.testingCache = new Testing("Java testing from CSV", new ArrayList<>());
```
Предлагаю сохранить csvResourceName  в поле класса
а всю логику перенести из конструктора.
Что бы она вызывалась из метода findAll()

MAMKUH6OPIIIUK replied:
Готово, теперь вся логика чтения ресурса в findAll().
Ещё добавил одно, возможно, спорное решение - findAll не кидает исключений, а просто возвращает пустой список, если ресурс имеет некорректный формат или его не удалось прочитать. Решил, что сервису внутреннюю кухню Dao знать особо ни к чему..

vlad-iv commented on spring-hw-01/src/main/java/ru/otus/spring/homework/oke/dao/TestingDaoCsv.java:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    @Override
+    public Testing findByThemeName(String themeName) {
```
Так как не будет кеширования, предлагаю убрать этот метод из интерфейса.

MAMKUH6OPIIIUK replied:
Готово, метод убрал. Сервис теперь использует findAll() метод Dao и самостоятельно выбирает из списка тестирование по нужной теме

MAMKUH6OPIIIUK committed:
Исправления по замечаниям первого круга проверки ДЗ:
1. Dao:
1.1. Тема тестирования перенесена в CSV-файл и теперь читается из него (не было в замечаниях, но думаю, что так будет логичнее)
1.2. Убрано кеширование, теперь ресурс читается при каждом обращении к объекту класса, а не при инициализации
1.3. Убран метод findByThemeName, теперь dao возвращает тестирования по всем темам, которые есть в CSV
2. Service:
2.1. Тема тестирования и проходной балл теперь являются полями класса и передаются при конфигурировании бина через контекст
2.2. Бин теперь конфигурируется через конструктор, т.к. конфигурирование через поля по информации из второго занятия - плохая практика
2.3. Метод тестирования студента теперь не имеет аргументов (и, соответственно, из Main класса запускается тоже без аргументов)
3. Unit test:
3.1. Заменил версии зависимостей для использования JUnit 5. Добавил конфигурацию maven-surefire-plugin
3.2. Переписал тест для использования аннотаций JUnit5

vlad-iv commented on spring-hw-01/src/main/java/ru/otus/spring/homework/oke/dao/TestingDaoCsv.java:
```diff
@@ -0,0 +1,206 @@
+     *
+     * @param csvResourceName имя ресурса внутри jar
+     */
+    public TestingDaoCsv(String csvResourceName) throws IOException, IncorrectCsvFormatException {
```
Можно удалить throws

MAMKUH6OPIIIUK replied:
Готово, сорри

vlad-iv commented on spring-hw-01/src/main/java/ru/otus/spring/homework/oke/Main.java:
```diff
@@ -0,0 +1,13 @@
+        ClassPathXmlApplicationContext context = new ClassPathXmlApplicationContext("/spring-context.xml");
+        TestingService testingService = context.getBean(TestingService.class);
+        testingService.executeStudentTesting();
+        context.close();
```
можно удалить закрытие

MAMKUH6OPIIIUK replied:
Убрал

vlad-iv commented on spring-hw-01/src/main/java/ru/otus/spring/homework/oke/dao/TestingDaoCsv.java:
```diff
@@ -0,0 +1,206 @@
+                */
+                return result;
+            }
+        } catch (IncorrectCsvFormatException | IOException e) {
```
Предлагаю все ошибки (которые еще не обернуты) обернуть в IncorrectCsvFormatException и выбросить его.

MAMKUH6OPIIIUK replied:
Обернул, теперь вместо пустого списка выбрасывается IncorrectCsvFormatException

vlad-iv commented on spring-hw-01/src/main/java/ru/otus/spring/homework/oke/dao/IncorrectCsvFormatException.java:
```diff
@@ -0,0 +1,11 @@
+package ru.otus.spring.homework.oke.dao;
+
+public class IncorrectCsvFormatException extends Exception {
```
Предлагаю наследоваться от RuntimeException  , это позволит убрать его из всех сигнатур методов

MAMKUH6OPIIIUK replied:
Готово. Из всех сигнатур убрал, в комментариях к методам оставил

vlad-iv commented on spring-hw-01/src/test/java/ru/otus/spring/homework/oke/service/SimpleTestingServiceTest.java:
```diff
@@ -0,0 +1,90 @@
+    @Test
+    public void test_executeStudentTesting_When_Theme_Not_Exists_Then_Not_Passed() {
+        final ByteArrayOutputStream myOut = new ByteArrayOutputStream();
+        System.setOut(new PrintStream(myOut));
```
Предлагаю использование в реализации IO консоли сделать явным.
Например:
Создать interface IOService с 2 методами printLine и readLine
Сделать его реализацию с System.in и System.out
а в тестах через конструктор их подменять на ByteArrayOutputStream

MAMKUH6OPIIIUK replied:
IO cервис реализовал, конфигурацию контекста поправил (sin/sout при помощи SpEL передаются в конструктор). Сервис тестирования студентов и юнит тесты теперь используют этот новый сервис

MAMKUH6OPIIIUK committed:
Исправления по результатам второй итерации проверки ДЗ
1. Кастомный Exception теперь наследуется от RuntimeException
2. Dao
2.1. IncorrectCsvFormatException убран из сигнатур всех методов (для javadoc описание оставил)
2.2. IncorrectCsvFormatException теперь выбрасывается при любых ошибках во время выполнения findAll(). В т.ч. потенциальный NPE при отсутствующем ресурсе\некорректном его имени теперь тоже будет приводить к IncorrectCsvFormatException, а не к возвращению пустого списка тестирований
2.3. throws убраны из сигнатуры конструктора
3. Main. Убрано закрытие контекста перед непосредственным завершением работы приложения
4. Реализован промежуточный IO сервис для работы с stdin\stdout через любые Input\Output Stream. В качестве бина добавляется в контекст через xml-конфиг. При инициализации бина sin и sout передаются в качестве аргументов конструктора при помощи SpEL (как иначе сделать - не придумал / не нагуглил)
5. TestingService. Теперь использует бин IO сервиса для вывода вопросов и вариантов ответов.
6. Юнит тесты. Теперь тоже используется IO сервис

MAMKUH6OPIIIUK committed:
Поправил формулировку в комментарии в соответствии с фактической реализацией

vlad-iv commented on spring-hw-01/src/main/java/ru/otus/spring/homework/oke/dao/TestingDaoCsv.java:
```diff
@@ -0,0 +1,202 @@
+            } else {
+                throw new IncorrectCsvFormatException("CSV resource not found");
+            }
+        } catch (IncorrectCsvFormatException e) {
```
Предлагаю не указывать этот кейс

MAMKUH6OPIIIUK replied:
Убрал перехват IncorrectCsvFormatException здесь

vlad-iv commented on spring-hw-01/src/main/java/ru/otus/spring/homework/oke/dao/TestingDaoCsv.java:
```diff
@@ -0,0 +1,202 @@
+            }
+        } catch (IncorrectCsvFormatException e) {
+            throw e;
+        } catch (Throwable e) {
```
Можно указать IOException

MAMKUH6OPIIIUK replied:
Готово

vlad-iv commented on spring-hw-01/src/main/java/ru/otus/spring/homework/oke/dao/TestingDaoCsv.java:
```diff
@@ -0,0 +1,202 @@
+        }
+    }
+
+    private void handleFormatException(Throwable e) {
```
Предлагаю заинлайнить метод.
Т.е. прописать catch для требуемых исключений

MAMKUH6OPIIIUK replied:
Готово. Заменил все вызовы метода catch-ами с пробросом IncorrectCsvFormatException.
Действительно, было не очень красиво

MAMKUH6OPIIIUK committed:
Исправления по результатам третьей итерации проверки ДЗ.
1. Dao
1.1. Убрал метод handleFormatException из кода, а в местах, где он вызывался, добавил catch с пробросом IncorrectCsvFormatException с соответствующим ситуации сообщением.
1.2. Переделал обработку IOException
